src/common/apply_permutation.h

```diff
@@ -0,0 +1,61 @@
+// Copyright (c) 2017, The Monero Project
+// 
+// All rights reserved.
+// 
+// Redistribution and use in source and binary forms, with or without modification, are
+// permitted provided that the following conditions are met:
+// 
+// 1. Redistributions of source code must retain the above copyright notice, this list of
+//    conditions and the following disclaimer.
+// 
+// 2. Redistributions in binary form must reproduce the above copyright notice, this list
+//    of conditions and the following disclaimer in the documentation and/or other
+//    materials provided with the distribution.
+// 
+// 3. Neither the name of the copyright holder nor the names of its contributors may be
+//    used to endorse or promote products derived from this software without specific
+//    prior written permission.
+// 
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+// MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL
+// THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+// STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
+// THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+// 
+// Most of this file is originally copyright (c) 2017 Raymond Chen, Microsoft
+// This algorithm is adapted from Raymond Chen's code:
+// https://blogs.msdn.microsoft.com/oldnewthing/20170109-00/?p=95145
+
+#include <vector>
+#include <functional>
+
+namespace tools
+{
+
+void apply_permutation(std::vector<size_t> permutation, const std::function<void(size_t, size_t)> &swap)
+{
+  for (size_t i = 0; i < permutation.size(); ++i)
+  {
+    size_t current = i;
+    while (i != permutation[current])
+    {
+      size_t next = permutation[current];
+      swap(current, next);
+      permutation[current] = current;
+      current = next;
+    }
+    permutation[current] = current;
+  }
+}
+
+template<typename T>
+void apply_permutation(const std::vector<size_t> &permutation, std::vector<T> &v)
```

Comment:
Why have the other overload? It doesn't appear to be used directly and is not taking advantage of a cpp.

Reply:
This is because it's the "typical" use, so it seemed useful to include it. What "advantage of a cpp" do you mean here ?

Reply:
Why have the `std::function` overload? Its definition is not in a cpp (currently), and its not being called directly AFAIK. So an argument could be made to drop that overload and do the swap logic directly in this function. There _could_ be some assembly output reduction as a result of the type-erasure, but the function is so short and there's still setup for the `std::function`, etc. 

In hindsight, I should not have mentioned `template<typename F>` above as that confused things a bit. To clarify, my recommendation would be to drop the `std::function` overload and move the definition here or to move the `std::function` overload implementation to a cpp file.